pkg/autoscaler/autoscaler.go

```diff
@@ -732,6 +518,96 @@ var _ = Describe("[Feature:Machines] Autoscaler should", func() {
+
+			By("Waiting for all Machines in the MachineSet to enter Running phase")
+			framework.WaitForMachineSet(client, transientMachineSet.GetName())
+
+			// To exercise the MaxNodesTotal mechanism we want to craft this
+			// MachineAutoscaler so that its maximum size would cause the cluster
+			// to grow beyond the MaxNodesTotal. This is calculated as the
+			// MaxNodesTotal - current number of nodes
+			// + 1 (for the initial size of the MachineSet)
+			// + 1 (to exceed the maximum)
+			By("Calculating maximum size for MachineAutoscaler")
+			nodes, err := framework.GetNodes(client)
+			Expect(err).NotTo(HaveOccurred())
+			maxMachineSetReplicas := int32(caMaxNodesTotal - len(nodes) + 2)
+
+			By(fmt.Sprintf("Creating a MachineAutoscaler backed by MachineSet %s - min: 1, max: %d",
+				transientMachineSet.GetName(), maxMachineSetReplicas))
+			asr := machineAutoscalerResource(transientMachineSet, 1, maxMachineSetReplicas)
+			Expect(client.Create(ctx, asr)).Should(Succeed())
+			cleanupObjects[asr.GetName()] = asr
+
+			// We want to create a workload that would cause the autoscaler to
+			// grow the cluster beyond the MaxNodesTotal. Since we have calculated
+			// the maximum MachineSet replicas to push beyond this boundary, we
+			// use that count as the replicas for our job.
+			jobReplicas := maxMachineSetReplicas
+			By(fmt.Sprintf("Creating scale-out workload: jobs: %v, memory: %s",
+				jobReplicas, workloadMemRequest.String()))
+			workload := newWorkLoad(jobReplicas, workloadMemRequest, targetedNodeLabel)
+			cleanupObjects[workload.GetName()] = workload
+			Expect(client.Create(ctx, workload)).Should(Succeed())
+
+			// At this point the autoscaler should be growing the cluster, we
+			// wait until the MachineSet has achieved its maximum replica size - 1.
+			expectedReplicas := maxMachineSetReplicas - 1
+			By(fmt.Sprintf("Waiting for MachineSet %s replicas to scale up", transientMachineSet.GetName()))
+			Eventually(func() (bool, error) {
+				current, err := framework.GetMachineSet(client, transientMachineSet.GetName())
+				if err != nil {
+					return false, err
+				}
+				if pointer.Int32PtrDerefOr(current.Spec.Replicas, 0) != expectedReplicas {
+					return false, nil
+				}
+				return true, nil
+			}, framework.WaitMedium, pollingInterval).Should(BeTrue())
+
+			// Now that the MachineSet has grown to the size we want, we check
+			// to ensure that it remains this size.
+			By(fmt.Sprintf("Watching MachineSet %s replicas to ensure they remain consistent", transientMachineSet.GetName()))
+			Consistently(func() (bool, error) {
+				current, err := framework.GetMachineSet(client, transientMachineSet.GetName())
+				if err != nil {
+					return false, err
+				}
+				if pointer.Int32PtrDerefOr(current.Spec.Replicas, 0) != expectedReplicas {
+					return false, nil
+				}
+				return true, nil
+			}, framework.WaitMedium, pollingInterval).Should(BeTrue())
```

Comment:
Should we not just watch the size of the cluster to make sure that it doesn't exceed that maximum count? Unless I've misunderstood, we are currently not actually checking that, just a derivative of that that implies the max size is observed, assuming our logic is correct?

I would have expected this to look something like

```
<create workload>
Eventually(<cluster hits max node count>)
Expect(<all machineset nodes are ready>) // To check there's no more coming, if there are, logic is borked
Consistently(<cluster stays at max node count>)
<delete workload>
``

Reply:
good point, i can switch this logic to focus around max cluster size instead.